di-ipv-credential-issuer-stub/src/main/java/uk/gov/di/ipv/stub/cred/service/ConfigService.java

```diff
@@ -0,0 +1,87 @@
+package uk.gov.di.ipv.stub.cred.service;
+
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.ssm.SsmClient;
+import software.amazon.awssdk.services.ssm.model.GetParametersByPathRequest;
+import software.amazon.awssdk.services.ssm.model.GetParametersByPathResponse;
+import software.amazon.awssdk.services.ssm.model.Parameter;
+import uk.gov.di.ipv.stub.cred.config.ClientConfig;
+import uk.gov.di.ipv.stub.cred.utils.StubSsmClient;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ConfigService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigService.class);
+    private static final long DEFAULT_CONFIG_CACHE_SECONDS = 60;
+    private static final Gson GSON = new Gson();
+    private static final SsmClient SSM_CLIENT = getSsmClient();
+    private static final Map<String, ClientConfig> CLIENT_CONFIGS = new HashMap<>();
+    private static final String CLIENT_CONFIG_BASE_PATH = "/stubs/credential-issuer-stub-clients";
+    private static final String ENVIRONMENT_ENV_VAR = "ENVIRONMENT";
+    private static final String TEST = "TEST";
+    private static final String CONFIG_CACHE_SECONDS_ENV_VAR = "CONFIG_CACHE_SECONDS";
+    private static Instant lastRefresh = null;
+
+    public static ClientConfig getClientConfig(String clientId) {
+        refreshClientConfigsIfRequired();
+
+        return CLIENT_CONFIGS.get(clientId);
+    }
+
+    public static Map<String, ClientConfig> getClientConfigs() {
+        refreshClientConfigsIfRequired();
+
+        return CLIENT_CONFIGS;
+    }
+
+    private static void refreshClientConfigsIfRequired() {
+        if (lastRefresh == null
+                || lastRefresh.isBefore(Instant.now().minusSeconds(getCacheDurationSeconds()))) {
+            LOGGER.info("Refreshing client configs");
+            getAllClientConfigs(null);
+            lastRefresh = Instant.now();
```

Comment:
This is not thread-safe - do you think it needs to be?

In the most common cases it will probably just mean the config gets refreshed twice, but in the worst case it could corrupt the hash map.

You could avoid the latter case by making `getAllClientConfigs` build a fresh map, rather than mutating the original.

Reply:
Hadn't thought about threading - even though I don't think we need it right now it's probably a good idea to deal with it in case we do in future.

I've not done loads of multithreaded stuff, but I'm tempted to make the `refreshClientConfigsIfRequried` method `synchronized`. I think that should allow multiple threads to use the same `lastRefresh` timestamp - which would be advantageous, as well as protect the map? I'll update `getAllClientConfigs` to generate a new map rather than mutate too though.

Interested to understand how the hash map might get corrupted in the current implementation though?

Reply:
Yeah, one neat thing about lambdas is that we usually don't have to worry about it (unless we deliberately write multithreaded code), but if this is running in a container and shared across requests, we probably do.

Making the method synchronised means that only one thread can enter the method at a time (and threads will just block until they can). In complex scenarios that can risk deadlocks etc. but that's shouldn't be a worry here.

The issue previously is that if multiple threads attempt to read/update the map at the same time, you can get race conditions in the internal workings of the hashmap. I'm a bit rusty on all the details, but they usually have 'buckets' of hashes, and as you add/remove items, these can change in size. If a thread reads/updates during one of these operations, it could potentially lead to items getting lost, getting stale/unexpected values, etc.

Reply:
Cool, makes sense. Thanks Joe